src/testing/slim_check/sampleable.lean

```diff
@@ -0,0 +1,268 @@
+
+```
+[2, 3, 3, 4, 4, 4, 5, 5, 6]
+```
+ -/
+def sampleable.lseq {α β γ} (f : α → β → γ) : lazy_list α → lazy_list β → lazy_list γ
+| lazy_list.nil xs := lazy_list.nil
+| (lazy_list.cons x xs) lazy_list.nil := lazy_list.nil
+| (lazy_list.cons x xs) ys := interleave (ys.map $ f x) (sampleable.lseq (xs ()) ys)
+
+/-- implementation of `sampleable nat` -/
+def nat.shrink' (k : ℕ) : ℕ → list ℕ → list ℕ
+| n ls :=
+if h : n ≤ 1
+  then ls.reverse
+  else
+    have 1 * n / 2 < n,
+      from nat.div_lt_of_lt_mul (nat.mul_lt_mul_of_pos_right (by norm_num) (by linarith)),
+    have n / 2 < n, by simpa,
+    let m := n / 2 in
+    nat.shrink' m ((k - m) :: ls)
+
+/-- `nat.shrink n` creates a list of smaller natural numbers by
+successively dividing by 2 and subtracting the difference from
+`n`. For example, `nat.shrink 100 = [50, 75, 88, 94, 97, 99]` -/
+def nat.shrink (n : ℕ) : list ℕ :=
+nat.shrink' n n []
+
+open gen
+
+instance sampleable_nat : sampleable ℕ :=
+{ sample := sized $ λ sz, fin.val <$> choose_any (fin $ succ (sz^3)) <|>
+                          fin.val <$> choose_any (fin $ succ sz),
+  shrink := lazy_list.of_list ∘ nat.shrink }
+
+/-- implementation of `sampleable int` -/
+def int.shrink' (k : ℕ) : ℕ → list ℤ → list ℤ
+| n ls :=
+if h : 1 < n
+  then
+    have n / 2 < n, from div_lt_self (by linarith) (by norm_num),
+    let m := n / 2 in
+    let m' := k - m in
+    int.shrink' m (m' :: -↑m' :: ls)
+  else ls.reverse
+
+/-- `int.shrink n` creates a list of integers by successively
+dividing by 2, subtracting the result from `n` and alternating the signs.
+For example, `int.shrink 40 = [-20, 20, -30, 30, -35, 35, -38, 38, -39, 39]` -/
+def int.shrink (i : ℤ) : list ℤ :=
+int.shrink' (int.nat_abs i) (int.nat_abs i) []
+
+instance sampleable_int : sampleable ℤ :=
+{ sample := sized $ λ sz,
+       let k := sz^5 in
+       (λ n : fin _, n.val - int.of_nat (k / 2) ) <$> choose_any (fin $ succ k),
+  shrink := lazy_list.of_list ∘ int.shrink   }
+
+instance sampleable_bool : sampleable bool :=
+{ sample := do { x ← choose_any bool,
+                 return x },
+  shrink := λ _, lazy_list.nil }
+
+instance sampleable_prod {β} [sampleable α] [sampleable β] : sampleable (α × β) :=
+{ sample := do { ⟨x⟩ ← uliftable.up $ sample α,
+                 ⟨y⟩ ← uliftable.up $ sample β,
+                 pure (x,y) },
+  shrink := λ x, sampleable.lseq prod.mk (shrink x.1) (shrink x.2) }
+
+/-- shrinking function for sum types -/
+def sum.shrink {β} [sampleable α] [sampleable β] : α ⊕ β → lazy_list (α ⊕ β)
+| (sum.inr x) := (shrink x).map sum.inr
+| (sum.inl x) := (shrink x).map sum.inl
+
+instance sampleable_sum {β} [sampleable α] [sampleable β] : sampleable (α ⊕ β) :=
+{ sample := uliftable.up_map sum.inl (sample α) <|>
+            uliftable.up_map sum.inr (sample β),
+  shrink := sum.shrink _ }
+
+instance sampleable_char : sampleable char :=
+{ sample := do { x ← choose_nat 0 3 dec_trivial,
+                 let str := " 0123abcABC:,;`\\/",
```

Comment:
How about something like this:
```
def sampleable_char (length : nat) (characters : string) : sampleable char :=
...

instance : sampleable char := sampleable_char 3 " 0123abcABC:,;`\\/"
```

@robertylewis Is sampleable the correct English spelling?

Reply:
I think "samplable" and "sampleable" are both acceptable. (Chrome's spell check doesn't like either.)

["samplable" dominated 20 years ago, but "sampleable" is rising fast](https://books.google.com/ngrams/graph?corpus=26&smoothing=3&content=sampleable%2C+samplable&year_end=2019&year_start=1800&direct_url=t1%3B%2Csampleable%3B%2Cc0%3B.t1%3B%2Csamplable%3B%2Cc0)

Reply:
@gebner What is the rationale behind the suggestion?

Reply:
It seemed potentially useful to customize the generated characters.  For example, if you want to do randomized testing on a JSON parser, you probably want `"{}:,\"truefals012."` instead.